### PR TITLE
CollectionAsync: use unique string keys to index _asyncTemplateItems (T1269855)

### PR DESCRIPTION
**Pull request by anna-shakhova**

### CODE REVIEW

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/treeView.async.tests.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                assert.ok($secondGroupCheckboxes.eq(0).hasClass(CHECKBOX_INDETERMINATE_CLASS), 'Second group root checkbox has indeterminate class');
+                assert.ok($secondGroupCheckboxes.eq(1).hasClass(CHECKBOX_INDETERMINATE_CLASS), 'Second group nested node checkbox has indeterminate class');
+                assert.notOk($secondGroupCheckboxes.eq(2).hasClass(CHECKBOX_CHECKED_CLASS), 'Second group leaf node 1 checkbox has not selected class');
+                assert.ok($secondGroupCheckboxes.eq(3).hasClass(CHECKBOX_CHECKED_CLASS), 'Second group leaf node 2 checkbox has selected class');
+
+                const $thirdRootNode = $treeRootNodes.eq(2);
+                const $thirdGroupCheckboxes = $thirdRootNode.find(`.${CHECKBOX_CLASS}`);
+
+                assert.notOk($thirdGroupCheckboxes.eq(0).hasClass(CHECKBOX_CHECKED_CLASS), 'Third group root checkbox has selected class');
+                assert.notOk($thirdGroupCheckboxes.eq(1).hasClass(CHECKBOX_CHECKED_CLASS), 'Third group nested node checkbox has selected class');
```
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/treeView.async.tests.js]:
```suggestion
                assert.notOk($thirdGroupCheckboxes.eq(1).hasClass(CHECKBOX_CHECKED_CLASS), 'Third group nested node checkbox has not selected class');
```

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/listParts/commonTests.js [hunk 2]
```diff
@@ -3683,6 +3683,54 @@ QUnit.module('regressions', moduleSetup, () => {
...
+            selectionMode: 'single',
+            selectedItemKeys: [data[0].id],
+        });
+
+        instance.option('_onItemsRendered', () => {
+            const listElement = instance.element();
+            const $firstGroup = $(listElement).find(`.${LIST_GROUP_CLASS}`).eq(0);
+            const $firstItemInFirstGroup = $firstGroup.find(`.${LIST_ITEM_CLASS}`).eq(0);
+
+            assert.ok($firstItemInFirstGroup.hasClass('dx-list-item-selected'), 'First item in first group should be selected');
```
Comment (EugeniyKiyashko) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets/listParts/commonTests.js]: Let's extract class to separate variable

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/treeView.async.tests.js (hunk 1 above)
Comment (anna-shakhova) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/treeView.async.tests.js]: fixed

File: packages/devextreme/testing/tests/DevExpress.ui.widgets/listParts/commonTests.js (hunk 2 above)
Comment (anna-shakhova) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets/listParts/commonTests.js]: fixed

File: packages/devextreme/js/__internal/ui/collection/m_collection_widget.base.ts [hunk 3]
```diff
@@ -876,7 +876,7 @@ const CollectionWidget = Widget.inherit({
     $element.attr(attributes);
   },
 
-  _renderItem(index, itemData, $container, $itemToReplace) {
+  _renderItem(index, itemData, $container, $itemToReplace, ...args) {
```
Comment (EugeniyKiyashko) [inline packages/devextreme/js/__internal/ui/collection/m_collection_widget.base.ts]: I assume this part is excess or I missed something

Comment (anna-shakhova) [reply packages/devextreme/js/__internal/ui/collection/m_collection_widget.base.ts]: extended arguments type and used uniqueKey explicitly

**Review by EugeniyKiyashko — APPROVED** [review]

**Review by nikkithelegendarypokemonster — APPROVED** [review]

**Pull request merged**